providers/class-two-factor-totp.php

```diff
@@ -131,14 +151,62 @@ public function user_two_factor_options( $user ) {
 		if ( empty( $key ) ) :
 			$key        = $this->generate_key();
 			$site_name  = get_bloginfo( 'name', 'display' );
+
+			// Must follow TOTP format for a "label":
+			// https://github.com/google/google-authenticator/wiki/Key-Uri-Format#label
+			// Do not URL encode, that will be done later.
 			$totp_title = apply_filters( 'two_factor_totp_title', $site_name . ':' . $user->user_login, $user );
+
+			$totp_url = add_query_arg(
+				array(
+					'secret' => rawurlencode( $key ),
+					'issuer' => rawurlencode( $site_name ),
+				),
+				'otpauth://totp/' . rawurlencode( $totp_title )
+			);
+
+			// Must follow TOTP format:
+			// https://github.com/google/google-authenticator/wiki/Key-Uri-Format
+			$totp_url = apply_filters( 'two_factor_totp_url', $totp_url, $user );
+			$totp_url = esc_url( $totp_url, array( 'otpauth' ) );
+
 			?>
+
 			<p>
 				<?php esc_html_e( 'Please scan the QR code or manually enter the key, then enter an authentication code from your app in order to complete setup.', 'two-factor' ); ?>
 			</p>
-			<p>
-				<img src="<?php echo esc_url( $this->get_google_qr_code( $totp_title, $key, $site_name ) ); ?>" id="two-factor-totp-qrcode" />
+			<p id="two-factor-qr-code">
+				<a href="<?php echo $totp_url; ?>">
```

Comment:
This needs to be escaped with `esc_url()`.

Reply:
It is escaped [above](https://github.com/WordPress/two-factor/blob/8e016214209739bd8b47994146e85727168bd898/providers/class-two-factor-totp.php#L171) (same w/ the other). Normally I'd escape as late as possible, but this one needs the extra param for the `otpauth` protocol, which would make those `echo`s fairly long and less readable, and duplicated instead of DRY.

I don't feel strongly either way though. I'm happy to change it if you'd prefer.